README.txt
= WDDX

http://wddx.rubyforge.org/

== DESCRIPTION

From http://www.openwddx.org/faq/:
 WDDX is an XML-based technology that enables the exchange of complex data between Web programming languages, 
 creating what some refer to as 'Web syndicate networks'. 
 WDDX consists of a language-independent representation of data based on an XML 1.0 DTD, and a set of 
 modules for a wide variety of languages that use WDDX. 
 WDDX can be used with HTTP, SMTP, POP, FTP and other Internet protocols that support transferring textual data.

WDDX Home: http://www.openwddx.org/

The Ruby WDDX Gem enables easy usage of WDDX in Ruby.

== Author

Copyright (C) 2005, 2006, 2007, 2008 by Stefan Saasen <s-NOSPAM-REMOVE@juretta.com> - http://juretta.com/

== Installation

Just run <tt>[sudo] gem install wddx</tt> to install the WDDX Gem.


== SYNOPSIS

=== Deserialization
You can very easily deserialize WDDX data from a file or from a string.

Simple usage example:
 
 require 'rubygems'
 require 'wddx'        
 # data.xml contains <?xml version="1.0">\
 # <wddxPacket version='1.0'><header><comment>A comment</comment></header><data><string>Klaus Tester</string></data></wddxPacket> 
 wddx = WDDX.load(File.open("data.xml"))
 puts wddx.comment # "A comment"
 puts wddx.data.class # String
 puts wddx.data # "Klaus Tester"
 
WDDX.load accepts a string with xml content, a file object or any object that provides a +read+ method.
 packet =  WDDX.load(open("http://wddx.rubyforge.org/wddx.xml"))
or
 packet =  WDDX.load(File.open("data.xml")) 
or
 packet =  WDDX.load(open("data.xml"))
or
 xml = "<wddxPacket version='1.0'><header/><data><number>123.667</number></data></wddxPacket>"
 packet = WDDX.load(xml)
 puts w.data # => 123.667
 
 xml = <<-EOF 
 <wddxPacket version='1.0'>
 <header/>
 <data>
     <struct>
         <var name='aNull'>
             <null/>
         </var>
         <var name='aString'>
             <string>a string</string>
         </var>
     </struct>
    </data>
  </wddxPacket> 
 EOF
 w = WDDX.load(xml)
 p w.data # => {"aString"=>"a string", "aNull"=>nil}
                                         
If the root element is a +struct+ you can use a shortcut version:
 
 w.data["aString"] # => "a string"

can be written as:

 w.aString # => "a string"


=== Serialization
The core Ruby Classes +Symbol+, +String+, +Numeric+, +true+, +false+, +nil+, +Hash+, +Array+ and +Time+ can be serialized by calling
+to_wddx+.

 require 'wddx'
 "Stefan Saasen".to_wddx        # => "<wddxPacket version='1.0'><header/><data>\
                                #     <string>Stefan Saasen</string></data></wddxPacket>"

 123.667.to_wddx                # => "<wddxPacket version='1.0'><header/><data>\
                                #     <number>123.667</number></data></wddxPacket>"

 Math::PI.to_wddx               # => "<wddxPacket version='1.0'><header/><data>\
                                #   <number>3.14159265358979</number></data></wddxPacket>"

 [1000, "Klaus Tester"].to_wddx # => "<wddxPacket version='1.0'><header/><data>\
                                #   <array length='2'><number>1000</number>\
                                #   <string>Klaus Tester</string></array></data></wddxPacket>"

=== WDDX.dump
You can use WDDX.dump to serialize Ruby objects.
 WDDX.dump("Hallo Welt") # => "<wddxPacket version='1.0'><header/><data><string>Hallo Welt</string></data></wddxPacket>"


Custom Ruby classes can be serialized by adding a +to_wddx_properties+ method (in the style of the YAML library) and calling WDDX.dump.

 class SerializeFromOutside
   def initialize(a,b,c)
     @a, @b, @c = a, b, c
   end
   
   def to_wddx_properties
     ["@a", "@c"]
   end  
 end 
 obj = SerializeFromOutside.new("This is a", "This is b", 123) 
 WDDX.dump(obj) # => "<wddxPacket version='1.0'><header/><data><struct><var name='a'>\
                # <string>This is a</string></var><var name='c'><number>123</number>\
                # </var></struct></data></wddxPacket>"

It is possible to include the WDDX module to get a behaviour similiar to the core Ruby classes.

 class MyObject
   include WDDX
   attr_accessor :name, :value, :price

   def to_wddx_properties
     ["@name", :custom_price]
   end               

   def custom_price
     @price * 1.05
   end
 end

 obj = MyObject.new("Stefan Saasen", 120)
 puts ob.to_wddx
 # => "<wddxPacket version='1.0'><header/><data><struct><var name='name'>\
 # <string>Stefan Saasen</string></var><var name='custom_price'><number>126.0</number>\
 # </var></struct></data></wddxPacket>"     
                                    
== WDDX Data types

WDDX defines some data types which can not be mapped to native ruby classes. WDDX::Binary represents a binary object (which is in fact a BASE64 encoded String). WDDX::RecordSet represents a RecordSet with data rows an column meta information.

== Type mapping
The following table shows the data type mapping WDDX <=> Ruby.
 WDDX Type      Ruby Type
 ---------      ---------------------
 String         String
 Number         Numeric
 Boolean        TrueClass, FalseClass
 Datetime       Time
 Null           nil
 Binary         WDDX::Binary
 Array          Array
 Struct         Hash
 Recordset      WDDX::RecordSet


== Rails Plugin

You can add the wddx gem to your rails application to be able to serialize ActiveRecord objects to WDDX.

In config/environment.rb add:
 require 'wddx'

Example:
 
 ...
 a_ar_obj.to_wddx # => <wddxPacket...


== See

Homepage:: http://rubyforge.org/projects/wddx/
Blog:: http://juretta.com/
Bugtracker:: http://rubyforge.org/tracker/?group_id=2715
WDDX:: http://www.openwddx.org/

== LICENSE:

(The MIT License)

Copyright (c) 2005, 2006, 2007, 2008 by Stefan Saasen <s-NOSPAM-REMOVE@juretta.com> - http://juretta.com/

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.4.0 / 2008-04-31
* Minor code cleanup.
* WDDX library now works in Ruby 1.8 and 1.9

=== 0.4.0 / 2008-03-04
* Added wddx binary. Pretty prints a wddx data structure.

  Usage: 
	wddx FILE|URL
  OR
    wddx < wddx.xml
  OR
    cat /path/to/wddx.xml | wddx

=== 0.3.3 / 2008-03-02
* Fixed test issues when running the test in a different timezone.

=== 0.3.2 / 2007-08-22
* Bugfix release: Fixes bug 13288.

=== 0.3.1 / 2007-04-19
* Bugfix release: Fixes bugs 10148 and 10149 reported by Rhett Sutphin
* Patch [#10152] kindly provided by Rhett Sutphin!

=== 0.3.0 / 2007-04-07
* New example wddx_serialize.rb added (see examples/ subdirectory).
* WDDX::Struct now includes Enumerable
* Documentation improved
* Symbol now responds_to?(:to_wddx)

=== 0.2.1 / 2006-12-20
Bugfix Release
* Fixed TcWddxComplexTest 

=== 0.2.0 / 2006-12-19
Added Rails support

Rails ActiveRecord::Base objects can be serialised to WDDX-XML using the +to_wddx+ method.

=== 0.1.0 / 2006-12-18
Initial release


test/test_helper.rb
# $Id: test_helper.rb 94 2006-12-24 13:05:36Z stefan $
require 'test/unit'
require File.dirname(__FILE__) + '/../lib/wddx'                                                   
FIXTURES = File.dirname(__FILE__) + "/fixtures"         

def read_fixture(f)
  File.read(File.join(FIXTURES, f)).gsub(/\t|\n/, "").gsub(/^(.*)<wddx/, "<wddx").gsub(/"/, "'")
end

test/tc_wddx_public_api_test.rb
#!/usr/bin/env ruby
#
#  Created by Stefan Saasen on 2006-12-17.
#  Copyright (c) 2006. All rights reserved.        

require File.dirname(__FILE__) + '/test_helper.rb' 
     

class MyObject
  include WDDX
  attr_accessor :name, :value, :price
  
  def initialize(name, price)
    @name, @price = name, price  
  end
  
  def to_wddx_properties
    ["@name", :custom_price]
  end               

  def custom_price
    @price * 1.05
  end
end

class AnotherObject
  include WDDX
  
  attr_accessor :my_object
  def to_wddx_properties
    ["@my_object"]
  end
end
    
class SerializeFromOutside
  
  def initialize(a,b,c)
    @a, @b, @c = a, b, c
  end
  
  def to_wddx_properties
    ["@a", "@c"]
  end  
  
end

# This is a TestCase containing UnitTests testing what is considered the public API.
# Here you can find usefule information and usage patterns.
class TcWddxPublicApiTest < Test::Unit::TestCase
  
  def test_simple_dump
    assert_equal("<wddxPacket version='1.0'><header/><data><string>Hallo Welt</string></data></wddxPacket>", WDDX.dump("Hallo Welt"))
  end
  
  def test_core_string_to_wddx
    assert_equal("<wddxPacket version='1.0'><header/><data><string>Hallo Welt</string></data></wddxPacket>", "Hallo Welt".to_wddx)
  end
  
  def test_core_number_to_wddx
    assert_equal("<wddxPacket version='1.0'><header/><data><number>10</number></data></wddxPacket>", 10.to_wddx)
    assert_equal("<wddxPacket version='1.0'><header/><data><number>-123</number></data></wddxPacket>", -123.to_wddx)
    assert_equal("<wddxPacket version='1.0'><header/><data><number>3.14159265358979</number></data></wddxPacket>", Math::PI.to_wddx)
  end
  
  def test_core_time_to_wddx
    assert_equal("<wddxPacket version='1.0'><header/><data><dateTime>2000-01-01T20:15:01Z</dateTime></data></wddxPacket>", Time.utc(2000,"jan",1,20,15,1).to_wddx)
  end 
  
  def test_dump
     vars = []
     vars << 23
     vars << Math::PI
     vars << {"key2" => Time.at(946702800)}
     assert_equal "<wddxPacket version='1.0'><header/><data><array length='3'><number>23</number><number>3.14159265358979</number><struct><var name='key2'><dateTime>2000-01-01T06:00:00+01:00</dateTime></var></struct></array></data></wddxPacket>", WDDX.dump(vars)
  end
  
  
  def test_load
    wddx = WDDX.load(File.open(FIXTURES + "/doc.1.xml"))
    assert_equal "A comment", wddx.comment
    assert_equal String, wddx.data.class
    assert_equal "Klaus Tester", wddx.data
  end
      
  def test_obj_serialization
    obj = MyObject.new("Stefan Saasen", 120)
    assert_equal "<wddxPacket version='1.0'><header/><data><struct><var name='name'><string>Stefan Saasen</string></var><var name='custom_price'><number>126.0</number></var></struct></data></wddxPacket>", obj.to_wddx    
    
    obj2 = AnotherObject.new
    obj2.my_object = obj
    assert_equal "<wddxPacket version='1.0'><header/><data><struct><var name='my_object'><struct><var name='name'><string>Stefan Saasen</string></var><var name='custom_price'><number>126.0</number></var></struct></var></struct></data></wddxPacket>", obj2.to_wddx                                                 
  end  
  
  
  def test_obj_serialization_without_include
    obj = SerializeFromOutside.new("This is a", "This is b", 123)
    assert_equal "<wddxPacket version='1.0'><header/><data><struct><var name='a'><string>This is a</string></var><var name='c'><number>123</number></var></struct></data></wddxPacket>", WDDX.dump(obj)
  end
  
  
  def test_shortcut_for_struct
    xml = "<wddxPacket version='1.0'><header/><data><struct><var name='name'><string>Stefan</string></var><var name='custom_price'><number>126.0</number></var></struct></data></wddxPacket>"
    packet = WDDX.load(xml)
        
    assert_equal("Stefan", packet.data["name"]) 
    assert_equal(126.0, packet.data["custom_price"])    
    
    assert_equal("Stefan", packet.name) 
    assert_equal(126.0, packet.custom_price) 
  end
  
  
end

test/tc_wddx_serializer_test.rb
#  $Id: tc_wddx_serializer_test.rb 109 2007-04-07 15:29:03Z stefan $
#  Created by Stefan Saasen.
#  Copyright (c) 2006. All rights reserved.

require File.dirname(__FILE__) + '/test_helper.rb'

class CustomClass
  attr_accessor :name
  def initialize(counter, name)
    @counter, @name = counter, name
  end
  
  def to_s
    "#{@counter}##{@name}"
  end
  
  def to_wddx_properties
    [:to_s, "@name"]
  end
end


class TcWddxSerializerTest < Test::Unit::TestCase

  def setup; super; end
  def teardown; super; end

=begin
  - aNull which is a null value,
  - aString which is the string 'a string',
  - aNumber which is the number -12.456,
  - aDateTime which is the date-time value June 12, 1998 4:32:12am,
  - aBoolean which is the boolean value true,
  - anArray which is an array of two elements (10 and 'second element'),
  - aBinary which contains 8 bytes of binary data encoded in base64,
  - anObject which is a structure with two properties s and n, and
  - aRecordset which is a recordset of two rows with fields NAME and AGE.  
=end
  # Test Basic data Types                                              
  def test_bdt_null
    xml = "<wddxPacket version='1.0'><header/><data><null/></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var nil
    assert_equal xml, wddx.to_xml
  end
  
  def test_bdt_string
    xml = "<wddxPacket version='1.0'><header/><data><string>Hallo Ruby</string></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var "Hallo Ruby"
    assert_equal xml, wddx.to_xml
  end
  
  def test_bdt_number
    xml = "<wddxPacket version='1.0'><header/><data><number>123</number></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var 123
    assert_equal xml, wddx.to_xml    
    
    xml = "<wddxPacket version='1.0'><header/><data><number>9.31231</number></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var 9.31231
    assert_equal xml, wddx.to_xml
    
    xml = "<wddxPacket version='1.0'><header/><data><number>-1231.23</number></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var(-1231.23)
    assert_equal xml, wddx.to_xml     
     
    xml = "<wddxPacket version='1.0'><header/><data><number>9.72398472938473e+24</number></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var 9723984729384729849231231
    assert_equal xml, wddx.to_xml      
  end
  
  def test_bdt_datetime
    xml = "<wddxPacket version='1.0'><header/><data><number>123</number></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var 123
    assert_equal xml, wddx.to_xml
  end
  
  def test_bdt_boolean
    xml = "<wddxPacket version='1.0'><header/><data><boolean value='true'/></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var true
    assert_equal xml, wddx.to_xml      
    
    xml = "<wddxPacket version='1.0'><header/><data><boolean value='false'/></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var false
    assert_equal xml, wddx.to_xml    
  end
  
  def test_bdt_array
    xml = "<wddxPacket version='1.0'><header/><data><array length='0'></array></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var []
    assert_equal xml, wddx.to_xml      
    
    xml = "<wddxPacket version='1.0'><header/><data><array length='3'><string>Hallo</string><string>Ruby</string><string>Welt</string></array></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var %w(Hallo Ruby Welt)
    assert_equal xml, wddx.to_xml                  
        
    xml = "<wddxPacket version='1.0'><header/><data><array length='3'><array length='3'><string>Hallo</string><string>Ruby</string><string>Welt</string></array><array length='3'><string>Hallo</string><string>Ruby</string><string>Welt</string></array><array length='3'><string>Hallo</string><string>Ruby</string><string>Welt</string></array></array></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var [%w(Hallo Ruby Welt), %w(Hallo Ruby Welt), %w(Hallo Ruby Welt)]
    assert_equal xml, wddx.to_xml    
    
    xml = "<wddxPacket version='1.0'><header/><data><array length='3'><array length='1'><array length='1'><array length='1'><number>1</number></array></array></array><array length='1'><number>2</number></array><array length='1'><number>3</number></array></array></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var [[[[1]]],[2],[3]]
    assert_equal xml, wddx.to_xml    
  end
  
  def test_bdt_binary
    xml = "<wddxPacket version='1.0'><header/><data><binary length='19'>U2VuZCByZWluZm9yY2VtZW50cw==</binary></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var WDDX::Binary.new("Send reinforcements")
    assert_equal xml, wddx.to_xml
    
    xml = "<wddxPacket version='1.0'><header/><data><binary length='19'>U2VuZCByZWluZm9yY2VtZW50cw==</binary></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var WDDX::Binary.new("Send reinforcements")
    assert_equal xml, wddx.to_xml         
    
    xml = "<wddxPacket version='1.0'><header/><data><binary length='19'>U2VuZCByZWluZm9yY2VtZW50cw==</binary></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    bin = WDDX::Binary.new
    bin.bin_data = "Send reinforcements"
    wddx.add_var bin
    assert_equal xml, wddx.to_xml
  end
  
  def test_bdt_object
    # Missing: include WddxSerializable Module to serialize arbitrary Objects
  end
  
  def test_bdt_recordset
    xml = "<wddxPacket version='1.0'><header/><data><recordset rowCount='2' fieldNames='NAME,AGE'><field name='NAME'><string>John Doe</string><string>Jane Doe</string></field><field name='AGE'><number>34</number><number>31</number></field></recordset></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var WDDX::RecordSet.new(['NAME', 'AGE'], [["John Doe", 34], ["Jane Doe", 31]])
    assert_equal xml, wddx.to_xml    
    
    xml = "<wddxPacket version='1.0'><header/><data><recordset rowCount='3' fieldNames='NAME,age'><field name='NAME'><string>John Doe</string><string>Jane Doe</string><string>Stefan Saasen</string></field><field name='age'><number>34</number><number>31</number><number>29</number></field></recordset></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var WDDX::RecordSet.new(['NAME', 'age'], [["John Doe", 34], ["Jane Doe", 31], ["Stefan Saasen", 29]])
    assert_equal xml, wddx.to_xml    
  end 
  
  
  def test_bdt_recordset2
    rows = []
    rows << [6, "MacBook Pro", 2000]
    rows << [10, "MacPro", 2200]
    rs = WDDX::RecordSet.new(["qty", "name", "price"], rows)
    assert_equal "<recordset rowCount='2' fieldNames='qty,name,price'><field name='qty'><number>6</number><number>10</number></field><field name='name'><string>MacBook Pro</string><string>MacPro</string></field><field name='price'><number>2000</number><number>2200</number></field></recordset>", rs.wddx_serialize
    wddx_packet = "<wddxPacket version='1.0'><header/><data><recordset rowCount='2' fieldNames='qty,name,price'><field name='qty'><number>6</number><number>10</number></field><field name='name'><string>MacBook Pro</string><string>MacPro</string></field><field name='price'><number>2000</number><number>2200</number></field></recordset></data></wddxPacket>" 
    assert_equal wddx_packet, WDDX.dump(rs)
    assert_equal wddx_packet, rs.to_wddx
  end
  
  def test_add_var
    xml = "<wddxPacket version='1.0'><header/><data><array length='0'></array></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var []
    assert_equal xml, wddx.to_xml
  end

  def test_ruby_hash
    xml = "<wddxPacket version='1.0'><header/><data><struct><var name='key1'><string>wert</string></var></struct></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var({"key1" => "wert"})
    assert_equal xml, wddx.to_xml
  end
  
  def test_serialize_simple
     wddx = "<wddxPacket version='1.0'><header><comment>Ruby rules!</comment></header><data><struct><var name='pi'><number>3.1415926</number></var><var name='cities'><array length='3'><string>Austin</string><string>Novato</string><string>Seattle</string></array></var></struct></data></wddxPacket>"
     
     @wddx = WDDX::Serializer::Serializer.new
     @wddx.add_comment "Ruby rules!"
     struct = WDDX::Struct.new
     struct["pi"] = 3.1415926
     struct["cities"] = ["Austin", "Novato", "Seattle"]
     
     @wddx.add_var struct
     assert_equal wddx, @wddx.to_xml
  end

  # Bugfix -> serialize control characters
  def test_serialize_all
     xml = "<wddxPacket version='1.0'><header/><data><array length='8'><string>This is array element #1</string><string>This is array element #2</string><string>This is array element #3</string><string>Some special characters, e.g., control characters <char code='0A'/><char code='0D'/>,pre-defined entities &lt;&amp;&gt;</string><struct><var name='B'><string>b</string></var><var name='A'><string>a</string></var></struct><string>true</string><number>-12.456</number><dateTime>2000-01-01T05:00:00Z</dateTime></array></data></wddxPacket>"
     wddx = WDDX::Serializer::Serializer.new
     struct = WDDX::Struct.new
     struct.put("B", "b")
     struct.put("A", "a")
     
     array = ["This is array element #1", "This is array element #2", "This is array element #3", "Some special characters, e.g., " +
                     "control characters " +
                     "\n\r," + 
                     "pre-defined entities <&>", struct, "true", -12.456, Time.at(946702800).gmtime]
     wddx.add_var array  
     assert_equal xml, wddx.to_xml     
  end
  
  # Bugifx -> serialize control characters \n und \r
  # 
  # End-of-line characters have platform and programming language specific 
  # representations. Different application environments may use either a 
  # single newline (0A), a single carriage return (0D), or a carriage return 
  # and newline combination (0D0A). For the purposes of successful data 
  # encoding and translation the elements <char code='0A'/> and 
  # <char code='0D'/> must be used to encode newline and carriage return 
  # characters when they should be preserved in the deserialized string.
  def test_serialize_special_characters
    xml = "<wddxPacket version='1.0'><header/><data><string><char code='0D'/></string></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var "\r"
    assert_equal xml, wddx.to_xml
    
    xml = "<wddxPacket version='1.0'><header/><data><string><char code='0A'/></string></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var "\n"
    assert_equal xml, wddx.to_xml
    
    xml = "<wddxPacket version='1.0'><header/><data><string><char code='0D'/><char code='0A'/></string></data></wddxPacket>"
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var "\r\n"
    assert_equal xml, wddx.to_xml        
  end
  
  # Calling add_var more than once will return an array element as the main container element for the serialized data.
  def test_serialize_consecutive_add_var_calls
    wddx = WDDX::Serializer::Serializer.new
    
    var1 = "A string"
    var2 = 1234
    var3 = WDDX::Binary.new("Bin data")

    wddx.add_var var1
    wddx.add_var var2
    wddx.add_var var3
    assert_equal(read_fixture("simple_array.2.xml"), wddx.to_xml)
  end
  
  def test_wddx_to_xml_result_cache
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var "A String"
    assert_equal("<wddxPacket version='1.0'><header/><data><string>A String</string></data></wddxPacket>", wddx.to_xml)
    
    wddx.add_var 123
    assert_equal("<wddxPacket version='1.0'><header/><data><array length='2'><string>A String</string><number>123</number></array></data></wddxPacket>", wddx.to_xml)
  end
  
  def test_to_s
    wddx = WDDX::Serializer::Serializer.new
    wddx.add_var "A String"
    assert_equal("<wddxPacket version='1.0'><header/><data><string>A String</string></data></wddxPacket>", wddx.to_s)
  end

  # Bugfix
  def test_serializer_to_wddx_properties
    xml = "<wddxPacket version='1.0'><header/><data><array length='3'><struct><var name='name'><string>test</string></var><var name='to_s'><string>1#test</string></var></struct><struct><var name='name'><string>test</string></var><var name='to_s'><string>2#test</string></var></struct><struct><var name='name'><string>test</string></var><var name='to_s'><string>3#test</string></var></struct></array></data></wddxPacket>"
    cls = (1..3).to_a.collect {|i| CustomClass.new(i, "test")}
    assert_equal xml, WDDX.dump(cls)
  end

end


lib/wddx/struct.rb
#  $Id: struct.rb 110 2007-04-19 15:36:06Z stefan $
#  Created by Stefan Saasen.
#  Copyright (c) 2006. All rights reserved.

module WDDX   

  # Struct is a sorted Hash. Use with care and only 
  # if the sort order is important - use Hash instead! 
  class Struct < WddxData  # Besser von Hash ableiten
    include Enumerable
    
    def initialize
      @order = []
      @hash = {}
    end  
    
    # Calls *block* once for each key in WDDX::Struct, passing the key and value as parameter.
    def each
      @order.each do |key|
         yield(key, @hash[key]) 
      end if block_given? 
    end
    alias :each_pair :each
    
    def <=>(other)
      raise NotImplementedError
    end
    
    def grep(pattern)
      @order.map{|key| @hash[key]}.grep pattern
    end
    
    # Calls *block* for each key in WDDX::Struct, passing the key as the parameter.
    def each_key(&block)
      @order.each &block
    end
    
    # Calls *block* for each key in WDDX::Struct, passing the value as the parameter.  
    def each_value
      @order.each do |key|
        yield @hash[key]
      end if block_given?
    end       
    
    # Returns true if the given key is present in WDDX::Struct   
    def key?(arg)
      @order.include?(arg)
    end
    alias :has_key? :key?
    alias :member? :key?
    alias :include? :key?
    
    # Returns true if WDDX::Struct contains no key/value pairs 
    def empty?
      @order.empty?
    end
    
    # Element assignment - Associates the value given by *value* with the key given by *key*
    def put(key, value)
      @hash[key] = value
      @order << key unless @order.include?(key)
    end
    alias :add :put

    # Element assignment - Associates the value given by *value* with the key given by *key*
    def []=(key, value)
      put(key, value)
    end
    
    # Element reference - Retrieves the *value* stored for *key*.                   
    def [](key)
      @hash[key]
    end
    
    # Returns the number of key/value pairs in WDDX::Struct
    def size
      @order.size
    end                
    alias :length :size
    
    # Returns the element at _index_. A negative index counts from the
    # end of _self_. Returns +nil+ if the index is out of range.
    def at(position)
      @hash[@order.at(position)]
    end
    
    def inspect
      map { |k, v| [k, v] }.inspect
    end
    
    # Removes all key/value pairs from WDDX::Struct
    def clear
      @order.clear
      @hash.clear
    end
    
    # Removes the key and its associated value from the struct
    def delete(key)
      @order.delete(key)
      @hash.delete(key)
    end
    
    # Serialize WDDX::Struct
    def wddx_serialize # :nodoc:
      xml = ["<struct>"]
      self.each do |k,v|
        xml << "<var name='#{k}'>"
        if v.respond_to?(:wddx_serialize)
          xml << v.wddx_serialize  
        else
          xml << nil.wddx_serialize  
        end
        xml << "</var>"
      end
      xml << "</struct>"
      xml.join('')
    end
  end # end Struct
end

lib/wddx/deserializer.rb
#  $Id: deserializer.rb 94 2006-12-24 13:05:36Z stefan $
#  Created by Stefan Saasen.
#  Copyright (c) 2006. All rights reserved.
 
require 'open-uri'
require 'rexml/document'
require 'rexml/streamlistener'
require File.dirname(__FILE__) + '/wddx_packet'
             
module WDDX
  module Deserializer # :nodoc:                       
    
    class Deserializer # :nodoc:    
      attr_reader :wddx_packet
      def initialize(xml_string)
        listener = Listener.new
        parser = REXML::Parsers::StreamParser.new(xml_string, listener)
        parser.parse
        @wddx_packet = listener.wddx_packet
      end
    end
    
    class Listener #:nodoc:
      include REXML::StreamListener       
      attr_reader :wddx_packet
      
      def initialize
        @wddx_packet = WddxPacket.new
        @stack = []
        @stack <<  Node.new("Root")
        @text = ""
      end
      
      def text(text)
        @text << text.dup #unless text.strip.empty?
      end
      
      def tag_start(name, attributes)
        node = Node.new(name, attributes)
        parent = @stack.last
        parent.children << node
        @stack << node       
      end

      def tag_end(name)
        #puts "Name: #{name}"
        case name.downcase
          when "comment"
            @wddx_packet.comment = @text.strip
          when "data"
            @wddx_packet.data = @current_node.to_ruby
          when "char" # empty tag
            char = @stack.pop
            @text << eval("0x"+char.attributes["code"]).chr
            return          
        end                                               
        @current_node = @stack.pop
        @current_node.text = @text.strip        
        @text = ""
      end
    end # class Listener    
    
    class Node #:nodoc:
      attr_reader :name, :attributes
      attr_accessor :children, :text
      def initialize(type, attributes = {})   
        raise ArgumentError, "Type is neccessary!" if type.nil?
        @children = []
        @attributes, @type = attributes, type
      end

      def has_children?
         !@children.empty?
      end 
      
      def to_ruby
         return case @type.downcase
           when "string"
             @text
           when "binary"
            bin = Binary.new
            bin.encoded_data = @text
            bin
           when "array", "field"
             @children.collect {|c| c.to_ruby }
           when "struct"
             h = {}
             @children.each {|c| var = c.to_ruby; h[var.key] = var.value}
             h                                 
           when "var"
             WDDX::Var.new(@attributes["name"], @children.first.to_ruby)
           when "null"
             nil     
           when "number"
             (@text =~ /\./ ) ? @text.to_f : @text.to_i
           when "boolean"
             (@attributes["value"].eql?("true")) ? true : false
           #  Date-time values-
           #  Date-time values are encoded according to the full form of ISO8601,
           #  e.g., 1998-9-15T09:05:32+4:0. Note that single-digit values for months,
           #  days, hours, minutes, or seconds do not need to be zero-prefixed. 
           #  While timezone information is optional, it must be successfully parsed
           #  and used to convert to local date-time values. Efforts should me made 
           #  to ensure that the internal representation of date-time values does not 
           #  suffer from Y2K problems and covers a sufficient range of dates. In 
           #  particular, years must always be represented with four digits.
           when "datetime"
             Time.xmlschema(@text) rescue Time.parse(@text)
           when "recordset"
             rs = WDDX::RecordSet.new(@attributes['fieldNames'].split(","))
             elems = []
             @children.collect{|col| elems << col.children }
             elems.transpose.each do |row|
               rs << row.collect {|c| c.to_ruby }
             end # .flatten.collect{|c| c.to_ruby }
             rs
           else
             raise RuntimeError, "#{type} not implemented!"     
             #"Missing -> #{@type}"
         end
      end
     
    end
  end        
  
end    


if __FILE__ == $0
  p WDDX::Deserializer::Deserializer.new(DATA.read).wddx_packet.data   
end

__END__
<?xml version="1.0"?><wddxPacket version='1.0'><header/><data><string>A</string></data></wddxPacket>